DOC-2254: Add Confluent Schema Registry schema migration via Shadowing (#1776)
redpanda-data/documentation: modules/manage/pages/disaster-recovery/shadowing/migrate-schemas-confluent.adoc

= Migrate Schemas from Confluent Schema Registry
:description: Replicate subjects, versions, and compatibility settings from a Confluent Schema Registry into a Redpanda shadow cluster.
:env-linux: true
:page-topic-type: how-to
:page-categories: Management, High Availability, Disaster Recovery
:learning-objective-1: Configure a shadow link that continuously replicates schemas from a Confluent Schema Registry
:learning-objective-2: Filter replication by context or subject and map source contexts to destination contexts
:learning-objective-3: Monitor schema replication status and resolve validation errors
// tag::single-source[]

// DRAFT-VERIFY (DOC-2254): This page documents an unreleased v26.2 feature (ENG-1171).
// Engineering PRs redpanda#30999 (HTTP source reader) and redpanda#30984 (failover) were
// still open when this draft was written (2026-07-02). Re-verify every DRAFT-VERIFY item
// against merged code and a running 26.2 build before publishing. SME: @pgellert.

ifndef::env-cloud[]
[NOTE]
====
include::shared:partial$enterprise-license.adoc[]
====
endif::[]

> Replicate glossterm:schema[,schemas] from a Confluent Schema Registry into a Redpanda glossterm:shadow cluster[] automatically and continuously. Schema replication through a glossterm:shadow link[] keeps subjects, versions, and compatibility settings synchronized with the source registry, so your applications keep working after a cutover without a separate schema migration step.

Author: ```suggestion
Redpanda allows you to replicate glossterm:schema[,schemas] from a Confluent Schema Registry into a Redpanda glossterm:shadow cluster[] automatically and continuously. Schema replication through a glossterm:shadow link[] keeps subjects, versions, and compatibility settings synchronized with the source registry, so your applications keep working after a cutover without a separate schema migration step.
```

:learning-objective-3: Monitor schema replication status and resolve validation errors
// tag::single-source[]

ifndef::env-cloud[]
[NOTE]
====
include::shared:partial$enterprise-license.adoc[]
====
endif::[]

When you migrate to Redpanda from a deployment that uses a Confluent Schema Registry, your producers and consumers depend on the glossterm:schema[,schemas] stored in that registry. Shadowing removes this migration obstacle: a glossterm:shadow link[] continuously replicates schemas from the source Confluent Schema Registry into the glossterm:Schema Registry[] built into the Redpanda glossterm:shadow cluster[], preserving glossterm:subject[] names, versions, and compatibility settings. Because both registries stay synchronized until cutover, your applications keep working on Redpanda without a separate schema migration step. Use this approach when you migrate from Confluent to Redpanda, or when you maintain a Redpanda disaster recovery cluster for a system that keeps its schemas in a Confluent Schema Registry.

After reading this page, you will be able to:

* [ ] {learning-objective-1}
* [ ] {learning-objective-2}
* [ ] {learning-objective-3}

> == How schema replication works

Reviewer: ```suggestion
== How HTTP API schema replication works
```
Need to distinguish between Confluent compatible SR replication and Redpanda

* Use *API mode* (`shadow_schema_registry_api`) when the source is a Confluent Schema Registry, or when you need to replicate only selected contexts or subjects, map source contexts to different destination contexts, or control how schemas that use unsupported features are handled with a validation policy.

Schema replication settings live in the shadow link configuration. Two cluster properties, `schema_registry_sync_memory_bytes` and `schema_registry_sync_parallelism`, tune how much memory and concurrency the shadow cluster uses while importing schemas. The defaults suit most deployments.
====

== Use cases

* *Migrate from Confluent to Redpanda*: Replicate schemas continuously while xref:manage:disaster-recovery/shadowing/setup.adoc[Shadowing] replicates your topic data, then cut applications over to Redpanda once both are in sync. No separate schema migration tooling is required.
* *Phased migration*: Use context and subject filters to migrate one team, application, or environment at a time.
* *Registry reorganization*: Map source contexts to different destination contexts to restructure your Schema Registry as part of the migration.

== Prerequisites

> * A shadow cluster running Redpanda version 26.2 or later. The schema replication feature activates after all brokers complete the upgrade.

Reviewer: ```suggestion
* A cluster running Redpanda version 26.2 or later. The schema replication feature activates after all brokers complete the upgrade.
```
It's just a cluster, doesn't need to be specified as a shadow cluster

* Network connectivity from the shadow cluster to the source Schema Registry HTTP endpoint.
* Credentials for the source registry with permission to read subjects, versions, and configuration. For Confluent Cloud, use a Schema Registry API key and secret.

> * A shadow link to the source cluster, or a plan to create one. See xref:manage:disaster-recovery/shadowing/setup.adoc[Configure Shadowing].

Reviewer: ```suggestion
* Basic Shadow link settings. See xref:manage:disaster-recovery/shadowing/setup.adoc[Configure Shadowing].
```

== Limitations

* HTTP basic authentication and mTLS are the supported authentication methods for the source registry.
* A shadow link replicates Schema Registry data in one mode only: either `shadow_schema_registry_topic` or `shadow_schema_registry_api`.
* Replication is one way, from the source registry to the shadow cluster. Destination contexts owned by the link are read-only until failover.
* Schemas that use Confluent features not supported by the Redpanda Schema Registry are not replicated as-is. Choose a <<choose-a-validation-policy,validation policy>> to control whether these schemas are skipped or imported without the unsupported fields.
* Topic data replication and schema replication are not coordinated with each other. Records serialized in the Confluent SerDes wire format can arrive on the shadow cluster before the schema IDs they contain have been replicated.

> 

Reviewer 2: Sorry I missed this from the first round, but let's please also add this limitation too:

```suggestion
* Deleting and recreating a subject: A hard delete on the source replicates on the next sync, not instantly. Wait until the subject is removed from the shadow cluster before registering a new schema under the same name, as recreating it too soon prevents the sync job from detecting the subject/subject-version change.

```

        ca_path: /path/to/ca.crt              # Optional: CA certificate for custom trust
    tail_interval: 10s                        # How often to poll for incremental changes
    full_sync_interval: 5m                    # How often to run a full scan
    max_source_requests_per_second: 30        # Rate limit for requests to the source registry
----

The intervals and rate limit are optional. If you omit them, Redpanda uses the defaults shown above.

To authenticate to the source registry with mTLS instead of HTTP basic authentication, omit `auth_options` and provide a client certificate and key in `tls_settings`.

=== Select contexts and subjects

By default, the link replicates the entire source registry. To replicate a subset, add a `source_filter` with the contexts or subjects to include:

[,yaml]
----
schema_registry_sync_options:
  shadow_schema_registry_api:
    # ...connection settings...
    source_filter:
      contexts:
      - "."                                   # The default context
      - ".prod"                               # A named context
      subjects:
      - orders-value                          # A subject in the default context
----

> The `contexts` and `subjects` lists accept literal names only. Wildcard and prefix patterns are not supported.

Reviewer: @pgellert how do the source filters on subjects work in conjunction with context filters? Does the example above really only apply "orders-value" subject in the default context? What about the .prod context?

Reviewer 2: It's explained here: https://github.com/redpanda-data/redpanda/blob/4ae658a0dea73bb117f458238f2e16c2c515b6fd/proto/redpanda/core/admin/v2/shadow_link.proto#L501-L517

The example above effectively turns into a filter for the default + prod contexts, as the subject filter part (orders-value) is a subset of the default context. So perhaps a better example would be removing the `.` context from this, and it's probably worth a paragraph explaining the behaviour here.

Author: Applied in 84f515f1 per Gellert's explanation: the example no longer overlaps (contexts: `.prod` in full, plus the single `orders-value` subject from the default context), and a new paragraph documents the union semantics from shadow_link.proto — contexts replicate entire contexts, subjects use qualified syntax (`:.staging:orders-value`), both lists union, and overlaps replicate once.

====

[[choose-a-validation-policy]]
=== Choose a validation policy

The `unsupported_schema_feature_policy` setting controls what happens when a source schema uses features that the Redpanda Schema Registry does not support. The unsupported Confluent Schema Registry features are:

* In schema definitions: rule sets and metadata tags.
* In subject configurations: override metadata, override rule sets, default metadata, default rule sets, and compatibility groups. Compatibility groups are not the same as compatibility levels, which do replicate.

The policy determines how the sync handles a schema or configuration that uses these features:

[cols="1,3"]
|===
| Policy | Behavior

| `FAIL` (default)
| The schema is not replicated. The sync records an error, reports it in the link status, and continues with the remaining schemas.

| `REMOVE`

> | The unsupported fields are removed and the rest of the schema is imported. The sync counts each modification in the link status.

Reviewer: ```suggestion
| The unsupported fields are removed and the rest of the schema is imported. The sync records each modification in the link status.
```

| Current sync
| The type of the sync in progress (`FULL` or `TAIL`) and the number of subject versions, compatibility configurations, and modes it has changed, including how many unsupported features were removed and how many errors occurred.

| Last full sync
| Start time, finish time, and change counts for the most recent completed full sync.

| Totals since task start
| Cumulative change and error counts since the schema replication task started.

| Last error
| The most recent replication error. With the `FAIL` validation policy, schemas that fail validation appear here.
|===

The schema replication task runs on the broker and shard that hosts the leader of the `_schemas` topic's partition. The counters in the status output are not persisted: expect them to reset to zero when that broker restarts or when leadership of the `_schemas` partition moves to another broker.

For detailed command options, see xref:reference:rpk/rpk-shadow/rpk-shadow-status.adoc[`rpk shadow status`]. For general link monitoring, see xref:manage:disaster-recovery/shadowing/monitor.adoc[Monitor Shadowing].

== Fail over

> Failing over shadow topics does not stop schema replication. As part of failover, pause the schema replication task by setting `paused: true` in the `schema_registry_sync_options` section of the link configuration. Pausing the task stops further syncing from the source registry and makes the write-blocked destination contexts writable, so your applications can register new schemas on the promoted cluster.

Reviewer: ```suggestion
For best results, shadow links should be failed over as a single unit, which prevents partial failover scenarios and unexpected results. Selective failover, such as topics only, does not stop schema replication. As part of failover, pause the schema replication task by setting `paused: true` in the `schema_registry_sync_options` section of the link configuration. Pausing the task stops further syncing from the source registry and makes the write-blocked destination contexts writable, so your applications can register new schemas on the promoted cluster.
```

* *Tail syncs* run frequently (default: every 10 seconds) to pick up incremental changes.
* *Full syncs* scan all selected subjects (default: every 5 minutes) to catch anything a tail sync missed.

Replicated schemas keep their original subject names and version IDs, so producers and consumers that reference schemas by ID continue to work after failover. Schemas that reference other schemas import in dependency order.

Before importing a schema, Redpanda validates it against the Redpanda Schema Registry implementation. If a schema uses features that Redpanda does not support, the sync either reports an error and skips the schema, or removes the unsupported fields and imports the rest, depending on the <<choose-a-validation-policy,validation policy>> you choose.

While the link is active, the destination contexts that the link replicates into are read-only: the shadow cluster rejects client writes to those contexts so that replicated schemas remain identical to the source. Contexts outside the link's filter remain writable.

[NOTE]
====

> This API-based mode is an alternative to the byte-for-byte `_schemas` topic replication described in xref:manage:disaster-recovery/shadowing/setup.adoc#schema-registry-synchronization[Configure Shadowing]. A shadow link uses one Schema Registry sync mode or the other, not both. Use the API-based mode when the source is a Confluent Schema Registry, or when you need to replicate only selected contexts or subjects.

Reviewer 3: I'm thinking if it would be worth to state when to use one instead of the other 🤔

Schema replication does not add any cluster configuration properties. All settings live in the shadow link configuration.
====

== Use cases

* *Migrate from Confluent to Redpanda*: Replicate schemas continuously while xref:manage:disaster-recovery/shadowing/setup.adoc[Shadowing] replicates your topic data, then cut applications over to Redpanda once both are in sync. No separate schema migration tooling is required.
* *Phased migration*: Use context and subject filters to migrate one team, application, or environment at a time.

> * *Registry reorganization*: Map source contexts to different destination contexts to restructure your Schema Registry as part of the migration.

Reviewer 3: @pgellert do we want to also mention here about disaster recovery? 🤔

Author: @bartoszpiekny-redpanda closing the loop on this one: the page's Use cases section covers the disaster-recovery scenario explicitly ("when you maintain a Redpanda disaster recovery cluster for a system that keeps its schemas in a Confluent Schema Registry"), and the intro's final sentence names both migration and DR as the two use cases. If you'd like DR called out more prominently than that, happy to adjust — otherwise I think this is covered.

// patterns are accepted before documenting either behavior.

== Configure schema replication

Add the `shadow_schema_registry_api` option to the `schema_registry_sync_options` section of your shadow link configuration file.

The following sample configuration file creates a shadow link that shadows all source topics and replicates schemas from a Confluent Schema Registry. The highlighted lines show the schema replication settings, which the sections that follow explain.

// NOTE: keep the lines= range pointing at the schema_registry_sync_options block
// if you edit this sample.
[,yaml,lines=20-38]
----
# Sample shadow link configuration with API-mode Schema Registry replication

name: confluent-migration                    # Unique name for this shadow link

client_options:
  bootstrap_servers:                         # Source Kafka cluster brokers
  - <source-broker-1>:<port>                 # Example: "pkc-xxxxx.us-east-1.aws.confluent.cloud:9092"
  - <source-broker-2>:<port>
  # For the TLS and authentication settings that the shadow cluster uses to
  # connect to the source Kafka cluster, see the complete configuration file
  # reference in Configure Shadowing.

> topic_metadata_sync_options:

Reviewer 3: I'm not sure we need to provide `topic_metadata_sync_options` in context of schema registry shadowing 🤔

Reviewer 2: Yes, we can remove the `topic_metadata_sync_options` section here, it's not needed.

Author: Removed in be454986 — the sample now goes straight from `client_options` to `schema_registry_sync_options`, and the highlight range is updated to match.

While the link is active, the destination contexts that the link replicates into are read-only: the shadow cluster rejects client writes to those contexts so that replicated schemas remain identical to the source. Contexts outside the link's filter remain writable.

[NOTE]
====
This API-based mode is an alternative to the byte-for-byte `_schemas` topic replication described in xref:manage:disaster-recovery/shadowing/setup.adoc#schema-registry-synchronization[Configure Shadowing]. A shadow link uses one Schema Registry sync mode or the other, not both:

* Use *topic mode* (`shadow_schema_registry_topic`) when the source is another Redpanda cluster and you want an exact, complete replica of its Schema Registry. Topic mode shadows the `_schemas` topic byte for byte, so it does not filter, remap, or validate schemas.
* Use *API mode* (`shadow_schema_registry_api`) when the source is a Confluent Schema Registry, or when you need to replicate only selected contexts or subjects, map source contexts to different destination contexts, or control how schemas that use unsupported features are handled with a validation policy.

> Schema replication does not add any cluster configuration properties. All settings live in the shadow link configuration.

Reviewer 2: This is false as stated, because we do add 2 new cluster configurations: https://github.com/redpanda-data/redpanda/blob/c153614a6972b0ee9e109ae93dc9d0a9369a1d37/src/v/config/configuration.cc#L3981-L3996

But they are escape hatches / tunables that won't commonly need to be adjusted.

Author: Fixed in be454986 — replaced the claim with the two tunables (`schema_registry_sync_memory_bytes`, `schema_registry_sync_parallelism`) and noted the defaults suit most deployments.

====

== Use cases

* *Migrate from Confluent to Redpanda*: Replicate schemas continuously while xref:manage:disaster-recovery/shadowing/setup.adoc[Shadowing] replicates your topic data, then cut applications over to Redpanda once both are in sync. No separate schema migration tooling is required.
* *Phased migration*: Use context and subject filters to migrate one team, application, or environment at a time.
* *Registry reorganization*: Map source contexts to different destination contexts to restructure your Schema Registry as part of the migration.

> == Prerequisites

Reviewer 2: One more to add here: if schema registry contexts are being used, customers must have the [schema_registry_enable_qualified_subjects](https://docs.redpanda.com/streaming/current/reference/properties/cluster-properties/#schema_registry_enable_qualified_subjects) config enabled -- https://docs.redpanda.com/streaming/current/manage/schema-reg/schema-reg-contexts/#prerequisites

Author: Added in be454986 as a prerequisite, with xrefs to the `schema_registry_enable_qualified_subjects` property reference and the Schema Registry Contexts prerequisites. Noted it's enabled by default.

* A shadow cluster running Redpanda version 26.2 or later. The schema replication feature activates after all brokers complete the upgrade.
* Network connectivity from the shadow cluster to the source Schema Registry HTTP endpoint.
* Credentials for the source registry with permission to read subjects, versions, and configuration. For Confluent Cloud, use a Schema Registry API key and secret.
* A shadow link to the source cluster, or a plan to create one. See xref:manage:disaster-recovery/shadowing/setup.adoc[Configure Shadowing].

// DRAFT-VERIFY (DOC-2254): Confirm with @pgellert whether the destination Schema Registry
// (_schemas topic) must be empty before enabling API-mode sync, or whether the destination

> // bootstrap in redpanda#30999 relaxes this. Topic mode requires it; API mode is unconfirmed.

Reviewer 2: In API mode, the schemas topic doesn't have to be empty, but the target contexts the link would import (scoped to the mapped contexts based on the `source_filter` and `destination` fields) need to be empty.

Ref: https://github.com/redpanda-data/redpanda/pull/31011

Author: Fixed in be454986 — the DRAFT-VERIFY is replaced with a prerequisite: the destination contexts the link replicates into (per `source_filter` and `destination` mapping) must be empty, and the rest of the shadow cluster's registry is unaffected.

> == Limitations

Reviewer 2: Another limitation to add is that there is currently no coordination between topic data replication and the schema registry syncing task, so topic data serialized with Confluent's SerDes format may contain schema IDs on the shadow cluster earlier than those schemas get replicated over.

Author: Added in be454986 as a limitation: topic data replication and schema replication are not coordinated, so records in the Confluent SerDes wire format can arrive before the schema IDs they contain have replicated.

Reviewer: @pgellert what's the impact of this? Users should not attempt to read from topics that rely on SR until SR synced? Or will this cause write or replication errors?

Reviewer 2: It's the former. We don't validate schema ids during topic partition data replication at the moment, so it wouldn't cause replication errors there.

Author: Extended in 84f515f1 with the impact per Gellert's answer: no replication errors (schema IDs aren't validated during topic data replication), but consumers that look up unreplicated schema IDs on the shadow cluster fail until the schemas arrive — so wait for schema replication to catch up before consuming schema-dependent shadow topics, with a pointer to Monitor replication status.

> * HTTP basic authentication is the only supported authentication method for the source registry.

Reviewer 2: mTLS is also supported as an authentication mechanism, in addition to HTTP Basic

Author: Fixed in be454986 — the limitation now lists HTTP basic and mTLS, and the connection section explains mTLS means a client certificate and key in `tls_settings` instead of `auth_options`.

* A shadow link replicates Schema Registry data in one mode only: either `shadow_schema_registry_topic` or `shadow_schema_registry_api`.
* Replication is one way, from the source registry to the shadow cluster. Destination contexts owned by the link are read-only until failover.
* Schemas that use Confluent features not supported by the Redpanda Schema Registry are not replicated as-is. Choose a <<choose-a-validation-policy,validation policy>> to control whether these schemas are skipped or imported without the unsupported fields.

// DRAFT-VERIFY (DOC-2254): Get the exact list of unsupported Confluent schema features

> // (rule sets? metadata tags?) from @pgellert and document them here.

Reviewer 2: In schema definitions:
- rule sets
- metadata tags
In configs:
- override metadata
- override rule set
- default metadata
- default rule set
- compatibility group (note: not the same as compatibility level)

Author: Documented in be454986 — the validation policy section now lists these (schema definitions: rule sets, metadata tags; configs: override/default metadata and rule sets, compatibility groups), with your note that compatibility groups are not compatibility levels.

// DRAFT-VERIFY (DOC-2254): ENG-1171 mentions wildcard support in filters, but the rpk
// source shows plain string lists for contexts/subjects. Confirm whether wildcards or

> // patterns are accepted before documenting either behavior.

Reviewer 2: Only plain string lists for contexts and subjects is supported

Author: Documented in be454986 — 'The `contexts` and `subjects` lists accept literal names only. Wildcard and prefix patterns are not supported.' DRAFT-VERIFY removed.

// DRAFT-VERIFY (DOC-2254): Defaults (10s / 5m / 30 rps) are from
// src/v/cluster_link/model/types.h on dev. Re-check at code-complete.

=== Select contexts and subjects

By default, the link replicates the entire source registry. To replicate a subset, add a `source_filter` with the contexts or subjects to include:

[,yaml]
----
schema_registry_sync_options:
  shadow_schema_registry_api:
    # ...connection settings...
    source_filter:
      contexts:
      - "."                                   # The default context
      - ".prod"                               # A named context
      subjects:
      - orders-value                          # A subject in the default context
----

Schema Registry contexts provide independent namespaces for subjects within one registry. The default context is named `.`. For more information, see xref:manage:schema-reg/schema-reg-contexts.adoc[Schema Registry Contexts].

=== Map source contexts to destination contexts

Choose how replicated contexts are named on the shadow cluster:

* `identity`: Keep the source context names (default behavior for migrations).

> * `exact`: Map each source context to a different destination context.

Reviewer 2: It's important to call out here that `exact` mapping needs to be exhaustive, and if the source schema registry has any contexts that are not mapped, that will lead to a failure of the schema syncing task.
We should recommend that people consider setting up a `source_scope` filter for the mapped source contexts if they expect that other contexts may later be created on the source schema registry that would fail the sync at a future inconvenient time.

Author: Added in be454986 as an IMPORTANT under the `exact` mapping example: mappings must cover every replicated context, an unmapped context fails the task, and we recommend scoping the `source_filter` contexts list to the mapped contexts to guard against contexts created later.

The `unsupported_schema_feature_policy` setting controls what happens when a source schema uses features that the Redpanda Schema Registry does not support:

[cols="1,3"]
|===
| Policy | Behavior

| `FAIL` (default)
| The schema is not replicated. The sync records an error, reports it in the link status, and continues with the remaining schemas.

| `REMOVE`
| The unsupported fields are removed and the rest of the schema is imported. The sync counts each modification in the link status.
|===

[,yaml]
----
schema_registry_sync_options:
  shadow_schema_registry_api:
    # ...connection settings, filters, and destination...
    unsupported_schema_feature_policy: FAIL
----

=== Create the shadow link

Create the shadow link with your completed configuration file:

[,bash]
----
rpk shadow create <link-name> --config-file shadow-config.yaml
----

For detailed command options, see xref:reference:rpk/rpk-shadow/rpk-shadow-create.adoc[`rpk shadow create`].

To change the schema replication settings on an existing link, see xref:reference:rpk/rpk-shadow/rpk-shadow-update.adoc[`rpk shadow update`].

// DRAFT-VERIFY (DOC-2254): Confirm the exact `rpk shadow create` argument order

> // (link name vs --config-file only) against the 26.2 rpk build.

Reviewer 2: Create doesn't take a link-name parameter:
```suggestion
[,bash]
----
rpk shadow create --config-file shadow-config.yaml
----

For detailed command options, see xref:reference:rpk/rpk-shadow/rpk-shadow-create.adoc[`rpk shadow create`].

To change the schema replication settings on an existing link, see xref:reference:rpk/rpk-shadow/rpk-shadow-update.adoc[`rpk shadow update`].
```

== Monitor replication status

Check schema replication progress and errors for a link:

[,bash]
----
rpk shadow status <link-name>
----

The Schema Registry section of the output reports:

[cols="1,3"]
|===
| Field | Description

| Inventory
| The number of selected subjects and subject versions on the source, compared with the number of subjects and versions on the destination. The registries are synchronized when the destination counts match the selected source counts. A destination that is behind the source indicates that replication is still in progress.

| Current sync
| The type of the sync in progress (`FULL` or `TAIL`) and the number of subject versions, compatibility configurations, and modes it has changed, including how many unsupported features were removed and how many errors occurred.

| Last full sync
| Start time, finish time, and change counts for the most recent completed full sync.

| Totals since task start
| Cumulative change and error counts since the schema replication task started.

> 

Reviewer 2: I am not sure where exactly to place this information on this page, but I think we should document somewhere that:
- These counters are ephemeral / not persisted, so it is expected that they reset to 0 on broker restarts or on partition movement of the `_schemas/0` topic partition.
- That the task is always running on the broker & shard that is the leader of the `_schemas` topic's partition.

Author: Added in be454986, at the end of the Monitor replication status section: the task runs on the broker/shard leading the `_schemas` partition, and the counters are not persisted — expect resets on broker restart or partition leadership movement.

| Last error
| The most recent replication error. With the `FAIL` validation policy, schemas that fail validation appear here.
|===

For detailed command options, see xref:reference:rpk/rpk-shadow/rpk-shadow-status.adoc[`rpk shadow status`]. For general link monitoring, see xref:manage:disaster-recovery/shadowing/monitor.adoc[Monitor Shadowing].

== Fail over

When you fail over a shadow link, schema replication from the source registry stops and the write-blocked destination contexts become writable, so your applications can register new schemas on the promoted cluster.

// DRAFT-VERIFY (DOC-2254): Failover behavior (single `paused` field pauses the sync task
// and lifts per-context write blocks) is from open PR redpanda#30984. Confirm the final
// mechanism and whether any manual steps apply to schema replication during failover.

> For the complete failover procedure, see xref:manage:disaster-recovery/shadowing/failover.adoc[Failover].

Reviewer 2: On failover, it is necessary to also pause the schema registry syncing task (through setting the paused field in the config to true) to stop any further syncing and to unblock the schema registry for writes.

As the draft verify note says, the relevant implementation is in https://github.com/redpanda-data/redpanda/pull/30984.

Author: Rewritten in be454986 — the Fail over section now states that failover alone does not stop schema replication, and that setting `paused: true` in `schema_registry_sync_options` is the step that stops syncing and unblocks the destination contexts for writes. DRAFT-VERIFY removed.
